labs/lab2/cord.cpp

```diff
@@ -48,36 +48,59 @@ cord_t cord_join(cord_t R, cord_t S) {
 }
 
 string cord_tostring(cord_t R) {
-    if (R == nullptr)
+
+    if (R == nullptr){
         return "";
-    if (R->left == nullptr && R->right == nullptr)
+    }
+
+    if (((R->left) == nullptr) && ((R->right) == nullptr)){
         return R->data;
-    else
-        return cord_tostring(R->left) + cord_tostring(R->right);
+
+    } else if (((R->left) != nullptr) && ((R->right) == nullptr)){
+        return cord_tostring(R->left);
+
+    } else if (((R->left) == nullptr) && ((R->right) != nullptr)){
+        return cord_tostring(R->right);
+
+    } else {
+        return (cord_tostring(R->left)+cord_tostring(R->right));
+    }
 }
 
 char cord_charat(cord_t R, size_t i)
 /*@requires 0 <= i && i < cord_length(R); @*/
 {
-    if (R->left == nullptr && R->right == nullptr)
+    if ((R->left == nullptr) && (R->right == nullptr)){
         return R->data[i];
-    if (i < R->left->len)
-        return cord_charat(R->left, i);
-    else
-        return cord_charat(R->right, i - R->left->len);
+    }
+    else if (i < R->left->len){
+        return cord_charat(R->left,i);
+    }
+    else{
+        return cord_charat(R->right,i-R->left->len);
+    }
 }
 
 cord_t cord_sub(cord_t R, size_t lo, size_t hi)
 /*@requires 0 <= lo && lo <= hi && hi <= cord_length(R); @*/
 {
-    if (lo == 0 && hi == R->len)
+    if (lo==0 && hi==cord_length(R)){
         return R;
-    if (R->left == nullptr && R->right == nullptr)
+    }
+
+    else if ((R->left != nullptr) || (R->right != nullptr)) {
+        if (hi < R->left->len) {
```

Comment:
Please notice `hi` is exclusive. When `hi == R->left->len`, you still need to `return cord_sub(R->left, lo, hi)`. However, your code will `return cord_join(cord_sub(R->left, lo, R->left->len), cord_sub(R->right, 0, hi - R->left->len))` in this case, which is equivalent to `return cord_join(cord_sub(R->left, lo, R->left->len), nullptr)`. This does not affect the result but will lead to an extra step.
```suggestion
        if (hi <= R->left->len) {
```